### ISSUE TITLE
Numerical issues in pyro.distributions.util.log_gamma

### ISSUE THREAD

**Issue by gokceneraslan** [uber/pyro#567]
`pyro.distributions.util.log_gamma` function is not so accurate for values close to zero and it returns `inf` for values <= `1e-8`. See the comparison of `torch.lgamma`, `pyro.distributions.util.log_gamma` and log gamma implementation from SciLua by @stepelu:

```python
import pyro
import torch
import numpy as np

def lgamma(z):
    gamma_r10 = 10.900511

    pi = torch.zeros_like(z)
    pi.fill_(np.pi)
    gamma_c = 2.0 * torch.sqrt(np.e/pi)

    # needs to be vectorized
    #if z < 0:
    #  return torch.log(pi) - torch.log(torch.abs(torch.sin(pi*z))) - lgamma(1.0 - z)

    sum = 2.48574089138753565546e-5
    sum += 1.05142378581721974210  / z
    sum += -3.45687097222016235469 / (z + 1.0)
    sum += 4.51227709466894823700  / (z + 2.0)
    sum += -2.98285225323576655721 / (z + 3.0)
    sum += 1.05639711577126713077  / (z + 4.0)
    sum += -1.95428773191645869583e-1 / (z + 5.0)
    sum += 1.70970543404441224307e-2  / (z + 6.0)
    sum += -5.71926117404305781283e-4 / (z + 7.0)
    sum += 4.63399473359905636708e-6  / (z + 8.0)
    sum += -2.71994908488607703910e-9 / (z + 9.0)

    # For z >= 0 gamma function is positive, no abs() required.
    return torch.log(gamma_c) + (z - 0.5)*torch.log(z  + gamma_r10 - 0.5) - (z - 0.5) + torch.log(sum)

for f in np.arange(7.0, 20.0):
    tf = torch.zeros(1)
    tf.fill_(f)
    n = torch.pow(10., -tf).float()

    print('Pyro  lgamma(1e-%i): ' % f, pyro.distributions.util.log_gamma(n).numpy()[0])
    print('Torch lgamma(1e-%i): ' % f, torch.lgamma(n).numpy()[0])
    print('Pugh  lgamma(1e-%i): ' % f, lgamma(n).numpy()[0])

for f in np.arange(7.0, 20.0):
    tf = torch.zeros(1)
    tf.fill_(f)
    n = torch.pow(10., tf).float()

    print('Pyro  lgamma(1e%i): ' % f, pyro.distributions.util.log_gamma(n).numpy()[0])
    print('Torch lgamma(1e%i): ' % f, torch.lgamma(n).numpy()[0])
    print('Pugh  lgamma(1e%i): ' % f, lgamma(n).numpy()[0])
```

Here is some parts from the output:
<details>

```
Pyro  lgamma(1e-7):  15.9424
Torch lgamma(1e-7):  16.1181
Pugh  lgamma(1e-7):  16.1181
Pyro  lgamma(1e-8):  inf
Torch lgamma(1e-8):  18.4207
Pugh  lgamma(1e-8):  18.4207
Pyro  lgamma(1e-9):  inf
Torch lgamma(1e-9):  20.7233
Pugh  lgamma(1e-9):  20.7233
Pyro  lgamma(1e-10):  inf
Torch lgamma(1e-10):  23.0259
Pugh  lgamma(1e-10):  23.0259
Pyro  lgamma(1e-11):  inf
Torch lgamma(1e-11):  25.3284
Pugh  lgamma(1e-11):  25.3284

... and goes on

Pyro  lgamma(1e7):  1.51181e+08
Torch lgamma(1e7):  1.51181e+08
Pugh  lgamma(1e7):  1.51181e+08
Pyro  lgamma(1e8):  1.74207e+09
Torch lgamma(1e8):  1.74207e+09
Pugh  lgamma(1e8):  1.74207e+09
Pyro  lgamma(1e9):  1.97233e+10
Torch lgamma(1e9):  1.97233e+10
Pugh  lgamma(1e9):  1.97233e+10
Pyro  lgamma(1e10):  2.20259e+11
Torch lgamma(1e10):  2.20259e+11
Pugh  lgamma(1e10):  2.20259e+11
Pyro  lgamma(1e11):  2.43284e+12
Torch lgamma(1e11):  2.43284e+12
Pugh  lgamma(1e11):  2.43284e+12
... goes on similarly
```
</details>
So does it make sense to replace pyro's log gamma implementation with this one?

**Comment by fritzo** [comment]
This sounds reasonable to me. It looks like SciLua is [MIT licensed](http://scilua.org/license.html), so I believe we just need to paste the copyright notice, maybe as a comment above the function?

**Comment by jpchen** [comment]
we should use the torch implementation. this implementation was temporarily (and hastily) written back when pytorch did not have `lgamma`.  there used to be a comment in the function noting to replace that when pytorch supports log gamma, log beta, etc

**Comment by neerajprad** [comment]
@jpchen - I think (and I may not have full context) the issue is that `torch.lgamma` does not yet support autograd, so we can't take gradients with respect to `z` using torch's lgamma. That said, I am not sure how numerically stable the gradients would be with our (or this) implementation, but this seems like an improvement anyhow.

**Comment by jpchen** [comment]
oh my mistake, from the comparison i assumed that torch now has a differentiable `lgamma` which would motivate the change. in that case the translation above of @stepelu's implementation in python looks like a good compromise until a differentiable implementation is merged upstream.  there seems to be other demand for this as well.

**Comment by gokceneraslan** [comment]
Since lgamma is in C++11 standard it was, I guess, easier to add it to PyTorch. However digamma, the derivative, is not in C++11. So ideally it must be implemented on PyTorch side in C/CUDA and defined as the backward step of lgamma. Until then, these implementations are required.

**Comment by stepelu** [comment]
@neerajprad , @gokceneraslan: exactly, the issue is that we don't have a digamma CUDA implementation in PyTorch, and the corresponding github issue is marked as low priority ;)
@fritzo yes, I wrote all of Scilua code under the MIT license so feel free to re-use it as you prefer

I would like to stress however that my main concern (in the forum thread from which this issue originated) was about the derivative of the approximation.
The numerical implementation I used in Scilua comes with numerical guarantees (see the PhD thesis mentioned in the comment). 
But computing the symbolic derivative of this approximation, which is what autodiff is doing numerically, does not result in corresponding bounds for the derivative of the gamma function.
That's why one would usually implement a separate numerical approximation for the derivative, i.e. the digamma function.
At least, a numerical investigation of the stability / accuracy of the current approach is warranted.

**Comment by gokceneraslan** [comment]
@stepelu I fully agree. This is why I use digamma from scipy in the backward step, slow but (hopefully) more accurate:

```python
class Lgamma(Function):
    @staticmethod
    def forward(self, input):
        self.save_for_backward(input)
        return torch.lgamma(input)

    @staticmethod
    def backward(self, grad_output):
        input, = self.saved_tensors
        res = torch.from_numpy(digamma(input.numpy())).type_as(input)
        return grad_output*res
```

I also wrote a small test to compare autodiff digamma and scipy digamma:

```python
import torch
from torch.autograd import Variable
import numpy as np
import scipy.special as sp

def lgamma(z):
    #same as above

torch.manual_seed(42)
x = Variable((torch.Tensor(6,6)).uniform_(1, 2), requires_grad=True)
y = lgamma(x)
y.backward(torch.ones(1))

print('X: \n', x.data.numpy())
d1 = x.grad.data.numpy()
print('SciLua autodiff digamma(X): \n', d1)
d2 = sp.digamma(x.data.numpy())
print('Scipy digamma(X): \n', d2)
print('Abs. difference: \n', np.abs(d1-d2))
print('Max abs. difference:', np.max(np.abs(d1-d2)))

```

For double highest abs. difference between scipy and scilua autodiff is 5.5e-15. whereas for float it's 4e-6. Is this reasonable or your concern was to see something worse?

**Comment by gokceneraslan** [comment]
Ah ok, I updated the range to `uniform_(0, 1e-10)` and now highest difference is `1e6` (not `1e-6`). So again for very small numbers autodiff of SciLua/Pugh approach is off.

**Comment by gokceneraslan** [comment]
How cephes digamma is added to eigen: https://bitbucket.org/eigen/eigen/pull-requests/150/add-digamma-for-cpu-cuda/diff

**Comment by eb8680** [comment]
@fritzo is there a new `log_gamma` implementation in PyTorch or distributions?  Can this be closed?

**Comment by fritzo** [comment]
Correct, `torch.distributions` now uses PyTorch's `lgamma`, and the upcoming `digamma` https://github.com/pytorch/pytorch/pull/3955 should make gradients more numerically stable. These will be available in Pyro 0.2 when we plan to switch to torch.distributions.
